### PR TITLE
[RFC] cgroup2: the first cut (yet another version)

### PR DESCRIPTION
**Pull request by AkihiroSuda**
alternative to #102

v2 package is completely split from v1.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

### CODE REVIEW

**Comment by crosbymichael** [comment]
Looks good.  How do you like this approach?  Do you think we should move forward with this implementation?  

I think this is nice, clean, and simple.

**Comment by AkihiroSuda** [comment]
@fuweid PTAL

**Comment by AkihiroSuda** [comment]
> Looks good. How do you like this approach? Do you think we should move forward with this implementation?

Let's try to move forward this and see if it work.

I have still a concern how we can support systemd, but I think the systemd interface can be revisited later, because CRI and Moby don't seem using it for v1.

For CRI and Moby, it seems we only need cgroupfs-based metrics functions: https://github.com/containerd/cgroups/issues/104

**Comment by AkihiroSuda** [comment]
Will `v%d` directory name cause some go mod issue? 😕 
https://blog.golang.org/using-go-modules

**Comment by fuweid** [comment]
> Will `v%d` directory name cause some go mod issue? 😕
> https://blog.golang.org/using-go-modules

I think we can add new go.mod to support v2 version.

File: v2/utils.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	var (
+		s = bufio.NewScanner(r)
+	)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return "", err
+		}
+		var (
+			text  = s.Text()
+			parts = strings.SplitN(text, ":", 3)
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply v2/utils.go]: doc seems wrong, where is such document?

**Comment by AkihiroSuda** [comment]
We have already several "v1" and "v2" pkgs in the main repo, probably we can keep this as-is?

**Comment by crosbymichael** [comment]
Do we want to create a v2 branch that we can work out of for this?

**Comment by AkihiroSuda** [comment]
As the package is separate out, unlikely to cause regression for the existing v1 code.
Probably no need to create a branch?

File: v2/state.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+package v2
+
+// State is a type that represents the state of the current cgroup
+type State string
+
+const (
+	Unknown  State = ""
+	Thawed   State = "thawed"
+	Frozen   State = "frozen"
+	Freezing State = "freezing"
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply v2/state.go]: Ah, "Freezing" state probably doesn't exist

File: v2/utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fuweid) [reply v2/utils.go]: I found that in https://www.kernel.org/doc/Documentation/cgroup-v2.txt

```
A cgroup which doesn't have any children or live processes can be
destroyed by removing the directory.  Note that a cgroup which doesn't
have any children and is associated only with zombie processes is
considered empty and can be removed::

  # rmdir $CGROUP_NAME

"/proc/$PID/cgroup" lists a process's cgroup membership.  If legacy
cgroup is in use in the system, this file may contain multiple lines,
one for each hierarchy.  The entry for cgroup v2 is always in the
format "0::$PATH"::

  # cat /proc/842/cgroup
  ...
  0::/test-cgroup/test-cgroup-nested
```

File: v2/utils.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		)
+		if len(parts) < 3 {
+			return "", fmt.Errorf("invalid cgroup entry: %q", text)
+		}
+		if parts[0] == "0" && parts[1] == "" {
```
Comment (fuweid) [inline v2/utils.go]: the parts[1] is empty means that ::. I just want it more clear :P

File: v2/utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply v2/utils.go]: Doc seems wrong

**Comment by AkihiroSuda** [comment]
addressed comments, but adding go.mod to `v2` seems difficult

```
~/gopath/src/github.com/containerd/cgroups/v2$ go mod tidy                                                                  
go: finding github.com/opencontainers/runtime-spec v1.0.1                                                                                  
go: finding github.com/containerd/cgroups latest                                                                                           
go: downloading github.com/opencontainers/runtime-spec v1.0.1                                                                              
go: downloading github.com/containerd/cgroups v0.0.0-20191011165608-5fbad35c2a7e                                                           
go: extracting github.com/containerd/cgroups v0.0.0-20191011165608-5fbad35c2a7e                                                            
go: extracting github.com/opencontainers/runtime-spec v1.0.1                                                                               
github.com/containerd/cgroups/v2 imports
        github.com/containerd/cgroups/stats/v2: module github.com/containerd/cgroups@latest found (v0.0.0-20191011165608-5fbad35c2a7e), but
 does not contain package github.com/containerd/cgroups/stats/v2
```

**Comment by crosbymichael** [comment]
just keep it simple and don't mess with go.mod if we don't have to.  getting the functionality in is what is important

**Comment by AkihiroSuda** [comment]
can we merge this as-is?

**Comment by fuweid** [comment]
@AkihiroSuda how about moving `stats/v2` into `v2/stats`? Yeah, the projects in whole org doesn't fully support go module. We do some work to prepare to migrate to go module. If we want `stats` out of cgroup controller functionality, we need 3 go.mod files in the future. One is for root dir of repo for v1, second one is for the stats API module, the last one is for v2 version. It seems not necessary. IMO, it is more clear to put `stats/v2` into `v2/stats`. 

And according to go.mod in v1 cgroup, it looks weird that we don't support go.mod for v2.

**Comment by AkihiroSuda** [comment]
Moved `stats/v2` to `v2/stats`.

Adding `go.mod` should be separate issue/PR.

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**